{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `Oligopool Calculator` in Action!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Author** Ayaan Hossain\n",
    "\n",
    "**Updated** Sepetember 30, 2024"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This `jupyter` notebook will demonstrate the usage of `Oligopool Calculator` and describe how to use its many functions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Table of Contents"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* [Notebook Setup](#Notebook-Setup)\n",
    "\n",
    "* [What is Oligopool Calculator?](#What-is-Oligopool-Calculator?)\n",
    "\n",
    "* [Design Mode Walthrough - An oligopool of 100K ribozyme variants](#Design-Mode-Walkthrough---An-oligopool-of-100K-ribozyme-variants)\n",
    "    * [Simulate the problem](#Simulate-the-problem)\n",
    "    * [Review constraints and formulate gameplan](#Review-constraints-and-formulate-gameplan)\n",
    "    * [Primer design review](#Primer-design-review)\n",
    "    * [Storing the plasmid background](#Storing-the-plasmid-background)\n",
    "    * [Designing the three primers](#Designing-the-three-primers)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notebook Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you [installed](https://github.com/ayaanhossain/oligopool#Installation) `Oligopool Calculator` successfully, you have everything you need to follow along."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's first import `oligopool`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import oligopool as op"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hopefully, the import worked without throwing up any errors! If you face issues importing, please [open an issue](https://github.com/ayaanhossain/oligopool/issues)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us also import `numpy` to do some math, and use `pandas` as our central data container."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy  as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will also need `matplotlib`/`seaborn` for plotting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lastly, we will need to use some operating system functions to manipulate files and directories."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We're all set."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What is Oligopool Calculator?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Oligonucleotide pools (oligopools) have revolutionized synthetic biology by enabling large-scale, cost-effective synthesis of hundreds of thousands of unique, short single-stranded DNA sequences through semiconductor-based technology, with major providers leading the market."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These oligopools facilitate the creation and study of various biological components, including de novo promoters, ribozymes, protein scaffolds, and CRISPR libraries. Massively parallel reporter assays (MPRAs) are utilized to clone and assess the functionality of these variants using high-throughput next-generation sequencing (NGS), which can generate billions of short reads per experiment."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Oligopool technology is shaping up to be a cornerstone of contemporary synthetic biology research."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looking ahead, advancements are expected in the number and length of oligos, NGS throughput, and reductions in manufacturing costs. Successful projects would then require not only core functional variants but also essential components like primers for amplification, spacers for structural integrity, and barcodes for accurate quantification via NGS."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![Oligopool Calculator Workflow](https://raw.githubusercontent.com/ayaanhossain/repfmt/refs/heads/main/oligopool/img/workflow.svg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Oligopool Calculator` is a comprehensive tool for designing and analyzing large-scale oligonucleotide pools used in synthetic biology research."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In **(a)** `Design Mode`, the algorithms optimize various elements of the oligopool, including barcodes, primers, spacers, and degenerate motifs. It ensures that these components are maximally distinct, non-repetitive, and free from undesired sequences or structures. The tool also enables the splitting of long oligos into shorter fragments with optimized overlaps for downstream assembly."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The **(b)** `Analysis Mode` of the Oligopool Calculator facilitates the processing of next-generation sequencing (NGS) data derived from barcoded amplicons. This mode allows for efficient mapping and counting of variants and their associated barcodes, producing count matrices for quantitative assessment of variant functionality. Two types of counting operations are supported: Association counting for assessing synthesis accuracy, and Combinatorial counting for analyzing multiple barcode combinations from sequencing reads."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overall `Oligopool Calculator` provides synthetic biologists with a unified platform for oligopool design and characterization, addressing the need for a standardized approach across different projects and labs. This tool eliminates the need for reinventing design scripts or using suboptimal strategies, making it particularly valuable for designing and analyzing pools containing millions of defined variants."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Design Mode Walkthrough - An oligopool of 6K ribozyme variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Suppose we want to design and test a library of 6,232 ribozyme variants based on the following architecture, which we characterized as part of our paper."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![Ribozyme Library Architecture](https://raw.githubusercontent.com/ayaanhossain/repfmt/refs/heads/main/oligopool/img/architecture.svg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the actual experiment, we designed them to be highly non-repetitive hammerhead ribozymes with a maximum shared repeat length of 16 bases using [`Non-Repetitive Parts Calculator`](https://github.com/ayaanhossain/nrpcalc), for simultaneous use as insulators in engineering large genetic system designs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To characterize the activity of our ribozyme library in _E. coli_ we adopted a two-barcode strategy for designing the oligopool. Because ribozymes are self-cleaving, the two barcodes, BC1 and BC2, placed before and after the core variant would be separated after catalysis. This allowed us to count the ratio of their separation to quantify the efficiency of the ribozymes on a proportional scale (i.e., 0 to 100% cleavage efficiency) and normalize against wildtype ribozyme cleavage efficiencies."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Simulate the problem"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the purposes of demonstration, let us first generate a random library of 59-73 mers (designing exact ribozymes is beyond the scope here) and assume that the library will be cloned into a 5kb plasmid backbone for characterization, which we will also generate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "rng = np.random.default_rng(seed=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna = ['A', 'T', 'G', 'C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "ribozymes = [''.join(rng.choice(dna, size=rng.integers(low=59, high=74))) for _ in range(6232)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA+0AAAEmCAYAAAAX93FNAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8hTgPZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA1X0lEQVR4nO3deViVdf7/8ddhRxJc2RQEJ1NcMreUbGZaCJgp09FLc9LScrJMyrRJx8olrUxTc1LL0a+hTZnZYlnO6LhXSu5LlqE5mn6VxQ3IDQg+vz/6eb6eXEA4cN9wno/ruq8L7s/n3Of95uDxfp17wWGMMQIAAAAAALbjZXUBAAAAAADg8gjtAAAAAADYFKEdAAAAAACbIrQDAAAAAGBThHYAAAAAAGyK0A4AAAAAgE0R2gEAAAAAsClCOwAAAAAANuVjdQF2UFxcrKNHj6pmzZpyOBxWlwMAAAAAqOaMMfrpp58UGRkpL68rH08ntEs6evSooqKirC4DAAAAAOBhDh8+rIYNG15xnNAuqWbNmpJ++WEFBwdbXA0AAAAAoLrLy8tTVFSUM49eCaFdcp4SHxwcTGgHAAAAAFSaki7R5kZ0AAAAAADYFKEdAAAAAACbIrQDAAAAAGBThHYAAAAAAGyK0A4AAAAAgE0R2gEAAAAAsClCOwAAAAAANsXfaQcAAACAauzQoUM6fvy41WVUmnr16ik6OtrqMtyG0A4AAAAA1dShQ4fUrFmczp07a3UplSYwsIa+/35PtQnuhHYAAAAAqKaOHz+uc+fOquPDYxQcEWN1ORUuL+OgNr71go4fP05oBwAAAABUDcERMaoT3dTqMlAG3IgOAAAAAACbsjS0FxUVadSoUYqNjVVgYKB+85vfaPz48TLGOOcYYzR69GhFREQoMDBQCQkJ2rdvn8t2Tp48qT59+ig4OFi1atXSgAEDdPr06cpuBwAAAAAAt7I0tE+cOFFvvvmmZsyYoT179mjixImaNGmSpk+f7pwzadIkvf7665o1a5Y2btyooKAgJSUl6fz58845ffr00bfffqsVK1bo888/1xdffKGBAwda0RIAAAAAAG5j6TXtGzZsUNeuXXX33XdLkmJiYvTee+9p06ZNkn45yj5t2jQ9//zz6tq1qyTp7bffVlhYmD755BP17t1be/bs0bJly7R582a1b99ekjR9+nT98Y9/1OTJkxUZGWlNcwAAAAAAlJOlR9pvueUWrVq1Snv37pUk7dy5U1999ZX+8Ic/SJIOHDigzMxMJSQkOB8TEhKijh07Ki0tTZKUlpamWrVqOQO7JCUkJMjLy0sbN26sxG4AAAAAAHAvS4+0/+1vf1NeXp6aNWsmb29vFRUV6aWXXlKfPn0kSZmZmZKksLAwl8eFhYU5xzIzMxUaGuoy7uPjozp16jjn/Fp+fr7y8/Od3+fl5bmtJwAAAAAA3MXSI+2LFi3Su+++qwULFmjbtm2aP3++Jk+erPnz51fo806YMEEhISHOJSoqqkKfDwAAAACAsrA0tD/zzDP629/+pt69e6tVq1Z64IEHNHToUE2YMEGSFB4eLknKyspyeVxWVpZzLDw8XNnZ2S7jP//8s06ePOmc82sjR45Ubm6uczl8+LC7WwMAAAAAoNwsDe1nz56Vl5drCd7e3iouLpYkxcbGKjw8XKtWrXKO5+XlaePGjYqPj5ckxcfHKycnR1u3bnXOWb16tYqLi9WxY8fLPq+/v7+Cg4NdFgAAAAAA7MbSa9q7dOmil156SdHR0WrRooW2b9+uqVOn6uGHH5YkORwOPfXUU3rxxRfVpEkTxcbGatSoUYqMjFS3bt0kSXFxcUpOTtYjjzyiWbNmqbCwUCkpKerduzd3jgcAAAAAVGmWhvbp06dr1KhRevzxx5Wdna3IyEg9+uijGj16tHPO8OHDdebMGQ0cOFA5OTm69dZbtWzZMgUEBDjnvPvuu0pJSdGdd94pLy8v9ejRQ6+//roVLQEAAAAA4DYOY4yxugir5eXlKSQkRLm5uZwqDwAAAKDa2LZtm9q1a6e7nktVneimVpdT4U4eSteKlx7S1q1b1bZtW6vLuarS5lBLr2kHAAAAAABXRmgHAAAAAMCmCO0AAAAAANgUoR0AAAAAAJsitAMAAAAAYFOEdgAAAAAAbIrQDgAAAACATRHaAQAAAACwKUI7AAAAAAA2RWgHAAAAAMCmCO0AAAAAANgUoR0AAAAAAJsitAMAAAAAYFOEdgAAAAAAbIrQDgAAAACATRHaAQAAAACwKUI7AAAAAAA2RWgHAAAAAMCmCO0AAAAAANgUoR0AAAAAAJsitAMAAAAAYFOEdgAAAAAAbIrQDgAAAACATRHaAQAAAACwKUI7AAAAAAA2RWgHAAAAAMCmCO0AAAAAANgUoR0AAAAAAJsitAMAAAAAYFOEdgAAAAAAbIrQDgAAAACATRHaAQAAAACwKUI7AAAAAAA2RWgHAAAAAMCmCO0AAAAAANgUoR0AAAAAAJsitAMAAAAAYFOEdgAAAAAAbIrQDgAAAACATRHaAQAAAACwKUI7AAAAAAA2RWgHAAAAAMCmLA/tR44cUd++fVW3bl0FBgaqVatW2rJli3PcGKPRo0crIiJCgYGBSkhI0L59+1y2cfLkSfXp00fBwcGqVauWBgwYoNOnT1d2KwAAAAAAuJWlof3UqVPq3LmzfH199e9//1vfffedpkyZotq1azvnTJo0Sa+//rpmzZqljRs3KigoSElJSTp//rxzTp8+ffTtt99qxYoV+vzzz/XFF19o4MCBVrQEAAAAAIDb+Fj55BMnTlRUVJRSU1Od62JjY51fG2M0bdo0Pf/88+ratask6e2331ZYWJg++eQT9e7dW3v27NGyZcu0efNmtW/fXpI0ffp0/fGPf9TkyZMVGRlZuU0BAAAAAOAmlh5pX7Jkidq3b6+ePXsqNDRUbdq00Zw5c5zjBw4cUGZmphISEpzrQkJC1LFjR6WlpUmS0tLSVKtWLWdgl6SEhAR5eXlp48aNl33e/Px85eXluSwAAAAAANiNpaH9v//9r9588001adJEy5cv16BBg/Tkk09q/vz5kqTMzExJUlhYmMvjwsLCnGOZmZkKDQ11Gffx8VGdOnWcc35twoQJCgkJcS5RUVHubg0AAAAAgHKzNLQXFxerbdu2evnll9WmTRsNHDhQjzzyiGbNmlWhzzty5Ejl5uY6l8OHD1fo8wEAAAAAUBaWhvaIiAg1b97cZV1cXJwOHTokSQoPD5ckZWVluczJyspyjoWHhys7O9tl/Oeff9bJkyedc37N399fwcHBLgsAAAAAAHZjaWjv3Lmz0tPTXdbt3btXjRo1kvTLTenCw8O1atUq53heXp42btyo+Ph4SVJ8fLxycnK0detW55zVq1eruLhYHTt2rIQuAAAAAACoGJbePX7o0KG65ZZb9PLLL6tXr17atGmTZs+erdmzZ0uSHA6HnnrqKb344otq0qSJYmNjNWrUKEVGRqpbt26Sfjkyn5yc7DytvrCwUCkpKerduzd3jgcAAAAAVGmWhvYOHTpo8eLFGjlypMaNG6fY2FhNmzZNffr0cc4ZPny4zpw5o4EDByonJ0e33nqrli1bpoCAAOecd999VykpKbrzzjvl5eWlHj166PXXX7eiJQAAAAAA3MbS0C5J99xzj+65554rjjscDo0bN07jxo274pw6depowYIFFVEeAAAAAACWsfSadgAAAAAAcGWEdgAAAAAAbIrQDgAAAACATRHaAQAAAACwKUI7AAAAAAA2RWgHAAAAAMCmCO0AAAAAANgUoR0AAAAAAJsitAMAAAAAYFOEdgAAAAAAbIrQDgAAAACATRHaAQAAAACwKUI7AAAAAAA2RWgHAAAAAMCmCO0AAAAAANgUoR0AAAAAAJsitAMAAAAAYFOEdgAAAAAAbKpMob1x48Y6ceLEJetzcnLUuHHjchcFAAAAAADKGNoPHjyooqKiS9bn5+fryJEj5S4KAAAAAABIPtcyecmSJc6vly9frpCQEOf3RUVFWrVqlWJiYtxWHAAAAAAAnuyaQnu3bt0kSQ6HQ/369XMZ8/X1VUxMjKZMmeK24gAAAAAA8GTXFNqLi4slSbGxsdq8ebPq1atXIUUBAAAAAIBrDO0XHDhwwN11AAAAAACAXylTaJekVatWadWqVcrOznYegb/grbfeKndhAAAAAAB4ujKF9hdeeEHjxo1T+/btFRERIYfD4e66AAAAAADweGUK7bNmzdK8efP0wAMPuLseAAAAAADw/5Xp77QXFBTolltucXctAAAAAADgImUK7X/5y1+0YMECd9cCAAAAAAAuUqbT48+fP6/Zs2dr5cqVuvHGG+Xr6+syPnXqVLcUBwAAAACAJytTaN+1a5duuukmSdLu3btdxrgpHQAAAAAA7lGm0L5mzRp31wEAAAAAAH6lTNe0AwAAAACAilemI+233377VU+DX716dZkLAgCgKjt06JCOHz9udRmVql69eoqOjra6DAAAqqUyhfYL17NfUFhYqB07dmj37t3q16+fO+oCAKDKOXTokJo1i9O5c2etLqVSBQbW0Pff7yG4AwBQAcoU2l977bXLrh87dqxOnz5droKAi3HECkBVcvz4cZ07d1YdHx6j4IgYq8upFHkZB7XxrRd0/Phx3ruAKor9LcDeyhTar6Rv3766+eabNXnyZHduFh6KI1b8RwJUVcERMaoT3dTqMgCgRJ66v+XvH6CPPvpQERERVpdS4fbs2WN1CSgnt4b2tLQ0BQQEuHOT8GAcsSK0AwCAiuWJ+1vH9u3UjkV/1z333GN1KZWqML/A6hJQRmUK7d27d3f53hijjIwMbdmyRaNGjXJLYcAFHLECANiJp51KzGnEnsGT9rfyMg5KMrrp/hGqH9vM6nIqXMY3adq9ZLZ+/vlnq0tBGZUptIeEhLh87+XlpaZNm2rcuHFKTEx0S2EAUB152s6+xA4/qhdPPJXYEy/b8qT3ak8+dfq60GiP+KDilw8pUJWVKbSnpqa6uw4AqPY8cWdf8swdflRfnnYqsSdetuWp79WcOg3YV7muad+6davz07kWLVqoTZs2bikKAKojT9vZlzxzhx+ewZNOJfY0nvZezanTgP2VKbRnZ2erd+/eWrt2rWrVqiVJysnJ0e23366FCxeqfv3617zNV155RSNHjtSQIUM0bdo0SdL58+f19NNPa+HChcrPz1dSUpLeeOMNhYWFOR936NAhDRo0SGvWrNF1112nfv36acKECfLxces99gDAbdjZB1DVeNIp1Bd69ZT3ak6dBuyvTMn2iSee0E8//aRvv/1WcXFxkqTvvvtO/fr105NPPqn33nvvmra3efNm/eMf/9CNN97osn7o0KFaunSpPvjgA4WEhCglJUXdu3fX+vXrJUlFRUW6++67FR4erg0bNigjI0MPPvigfH199fLLL5elNQAAAPx/53JPSHKob9++VpdS6ThdHIBdlCm0L1u2TCtXrnQGdklq3ry5Zs6cec03ojt9+rT69OmjOXPm6MUXX3Suz83N1dy5c7VgwQLdcccdkn65lj4uLk5ff/21OnXqpP/85z/67rvvtHLlSoWFhemmm27S+PHjNWLECI0dO1Z+fn5laQ8AAACSCs/+JE+6y7bE6eIA7KdMob24uFi+vr6XrPf19VVxcfE1bWvw4MG6++67lZCQ4BLat27dqsLCQiUkJDjXNWvWTNHR0UpLS1OnTp2UlpamVq1auZwun5SUpEGDBunbb7+94jX2+fn5ys/Pd36fl5d3TTVbibuZegZP6p07iwOA/XnKXbYlThcHYD9lCu133HGHhgwZovfee0+RkZGSpCNHjmjo0KG68847S72dhQsXatu2bdq8efMlY5mZmfLz83NeM39BWFiYMjMznXMuDuwXxi+MXcmECRP0wgsvlLpOu+BuptWfJ56G6O8foI8++lARERFWl1LhPOnDGAAAALhHmUL7jBkzdO+99yomJkZRUVGSpMOHD6tly5Z65513SrWNw4cPa8iQIVqxYoUCAgLKUkaZjRw5UsOGDXN+n5eX5+zDzribafXnaachHtu3UzsW/V333HOP1aVUKk/6IAoAAADlU6bQHhUVpW3btmnlypX6/vvvJUlxcXEup7KXZOvWrcrOzlbbtm2d64qKivTFF19oxowZWr58uQoKCpSTk+NytD0rK0vh4eGSpPDwcG3atMllu1lZWc6xK/H395e/v3+pa7Ub7mZa/XnKaYi/vMae8yGFJ34QBQAAgPK5ptC+evVqpaSk6Ouvv1ZwcLDuuusu3XXXXZJ+uXFcixYtNGvWLP32t78tcVt33nmnvvnmG5d1Dz30kJo1a6YRI0YoKipKvr6+WrVqlXr06CFJSk9P16FDhxQfHy9Jio+P10svvaTs7GyFhoZKklasWKHg4GA1b978WloDYCHP+pACAAAAKL1rCu3Tpk3TI488ouDg4EvGQkJC9Oijj2rq1KmlCu01a9ZUy5YtXdYFBQWpbt26zvUDBgzQsGHDVKdOHQUHB+uJJ55QfHy8OnXqJElKTExU8+bN9cADD2jSpEnKzMzU888/r8GDB1fpI+kAAFQ1nnLPBk/pEwBgH9cU2nfu3KmJEydecTwxMVGTJ08ud1EXvPbaa/Ly8lKPHj2Un5+vpKQkvfHGG85xb29vff755xo0aJDi4+MVFBSkfv36ady4cW6rAQBQfp4SdDylz4t54g00Je5NAQCoPNcU2rOysi77p96cG/Px0bFjx8pczNq1a12+DwgI0MyZMzVz5swrPqZRo0b617/+VebnBABUHAJd9edpN9Dk3hQAgMp2TaG9QYMG2r17t66//vrLju/atcsj/mwTAKB0CHSeg3tTAABQMa4ptP/xj3/UqFGjlJycfMmfaTt37pzGjBnjcX+6CQBQMgIdAABA2VxTaH/++ef18ccf64YbblBKSoqaNv1lB+z777/XzJkzVVRUpOeee65CCgUAAAAAwNNcU2gPCwvThg0bNGjQII0cOVLGGEmSw+FQUlKSZs6cqbCwsAopFAAAAAAAT3NNoV36vxu/nTp1Sj/88IOMMWrSpIlq165dEfUBAAAAAOCxrjm0X1C7dm116NDBnbUAAAAAAICLeFldAAAAAAAAuDxCOwAAAAAANkVoBwAAAADApgjtAAAAAADYFKEdAAAAAACbIrQDAAAAAGBThHYAAAAAAGyK0A4AAAAAgE0R2gEAAAAAsClCOwAAAAAANkVoBwAAAADApgjtAAAAAADYFKEdAAAAAACbIrQDAAAAAGBThHYAAAAAAGyK0A4AAAAAgE0R2gEAAAAAsClCOwAAAAAANkVoBwAAAADApgjtAAAAAADYFKEdAAAAAACbIrQDAAAAAGBThHYAAAAAAGyK0A4AAAAAgE0R2gEAAAAAsClCOwAAAAAANkVoBwAAAADApgjtAAAAAADYFKEdAAAAAACbIrQDAAAAAGBThHYAAAAAAGyK0A4AAAAAgE0R2gEAAAAAsClCOwAAAAAANkVoBwAAAADApgjtAAAAAADYlKWhfcKECerQoYNq1qyp0NBQdevWTenp6S5zzp8/r8GDB6tu3bq67rrr1KNHD2VlZbnMOXTokO6++27VqFFDoaGheuaZZ/Tzzz9XZisAAAAAALidpaF93bp1Gjx4sL7++mutWLFChYWFSkxM1JkzZ5xzhg4dqs8++0wffPCB1q1bp6NHj6p79+7O8aKiIt19990qKCjQhg0bNH/+fM2bN0+jR4+2oiUAAAAAANzGx8onX7Zsmcv38+bNU2hoqLZu3arf/e53ys3N1dy5c7VgwQLdcccdkqTU1FTFxcXp66+/VqdOnfSf//xH3333nVauXKmwsDDddNNNGj9+vEaMGKGxY8fKz8/PitYAAAAAACg3W13TnpubK0mqU6eOJGnr1q0qLCxUQkKCc06zZs0UHR2ttLQ0SVJaWppatWqlsLAw55ykpCTl5eXp22+/vezz5OfnKy8vz2UBAAAAAMBubBPai4uL9dRTT6lz585q2bKlJCkzM1N+fn6qVauWy9ywsDBlZmY651wc2C+MXxi7nAkTJigkJMS5REVFubkbAAAAAADKzzahffDgwdq9e7cWLlxY4c81cuRI5ebmOpfDhw9X+HMCAAAAAHCtLL2m/YKUlBR9/vnn+uKLL9SwYUPn+vDwcBUUFCgnJ8flaHtWVpbCw8OdczZt2uSyvQt3l78w59f8/f3l7+/v5i4AAAAAAHAvS4+0G2OUkpKixYsXa/Xq1YqNjXUZb9eunXx9fbVq1SrnuvT0dB06dEjx8fGSpPj4eH3zzTfKzs52zlmxYoWCg4PVvHnzymkEAAAAAIAKYOmR9sGDB2vBggX69NNPVbNmTec16CEhIQoMDFRISIgGDBigYcOGqU6dOgoODtYTTzyh+Ph4derUSZKUmJio5s2b64EHHtCkSZOUmZmp559/XoMHD+ZoOgAAAACgSrM0tL/55puSpNtuu81lfWpqqvr37y9Jeu211+Tl5aUePXooPz9fSUlJeuONN5xzvb299fnnn2vQoEGKj49XUFCQ+vXrp3HjxlVWGwAAAAAAVAhLQ7sxpsQ5AQEBmjlzpmbOnHnFOY0aNdK//vUvd5YGAAAAAIDlbHP3eAAAAAAA4IrQDgAAAACATRHaAQAAAACwKUI7AAAAAAA2RWgHAAAAAMCmCO0AAAAAANgUoR0AAAAAAJsitAMAAAAAYFOEdgAAAAAAbIrQDgAAAACATRHaAQAAAACwKUI7AAAAAAA2RWgHAAAAAMCmCO0AAAAAANgUoR0AAAAAAJsitAMAAAAAYFOEdgAAAAAAbIrQDgAAAACATRHaAQAAAACwKUI7AAAAAAA2RWgHAAAAAMCmCO0AAAAAANgUoR0AAAAAAJsitAMAAAAAYFOEdgAAAAAAbIrQDgAAAACATRHaAQAAAACwKUI7AAAAAAA2RWgHAAAAAMCmCO0AAAAAANgUoR0AAAAAAJsitAMAAAAAYFOEdgAAAAAAbIrQDgAAAACATRHaAQAAAACwKUI7AAAAAAA2RWgHAAAAAMCmCO0AAAAAANgUoR0AAAAAAJsitAMAAAAAYFOEdgAAAAAAbIrQDgAAAACATVWb0D5z5kzFxMQoICBAHTt21KZNm6wuCQAAAACAcqkWof3999/XsGHDNGbMGG3btk2tW7dWUlKSsrOzrS4NAAAAAIAyqxahferUqXrkkUf00EMPqXnz5po1a5Zq1Kiht956y+rSAAAAAAAoMx+rCyivgoICbd26VSNHjnSu8/LyUkJCgtLS0i77mPz8fOXn5zu/z83NlSTl5eVVbLHldPr0aUnSyR/T9XP+OYurqXh5GT9KknKP7JOvj8PiaiqHp/VMv9Wfp/Xsaf1Kntcz/VZ/ntazp/UreV7PHtdv5iFJv2Qnu+e7C/UZY646z2FKmmFzR48eVYMGDbRhwwbFx8c71w8fPlzr1q3Txo0bL3nM2LFj9cILL1RmmQAAAAAAXOLw4cNq2LDhFcer/JH2shg5cqSGDRvm/L64uFgnT55U3bp15XDY99OnvLw8RUVF6fDhwwoODra6nArnaf1Kntcz/VZ/ntazp/UreV7P9Fv9eVrPntav5Hk90699GWP0008/KTIy8qrzqnxor1evnry9vZWVleWyPisrS+Hh4Zd9jL+/v/z9/V3W1apVq6JKdLvg4GDb/wK6k6f1K3lez/Rb/Xlaz57Wr+R5PdNv9edpPXtav5Ln9Uy/9hQSElLinCp/Izo/Pz+1a9dOq1atcq4rLi7WqlWrXE6XBwAAAACgqqnyR9oladiwYerXr5/at2+vm2++WdOmTdOZM2f00EMPWV0aAAAAAABlVi1C+3333adjx45p9OjRyszM1E033aRly5YpLCzM6tLcyt/fX2PGjLnk1P7qytP6lTyvZ/qt/jytZ0/rV/K8num3+vO0nj2tX8nzeqbfqq/K3z0eAAAAAIDqqspf0w4AAAAAQHVFaAcAAAAAwKYI7QAAAAAA2BShHQAAAAAAmyK028zYsWPlcDhclmbNmjnH9+/frz/96U+qX7++goOD1atXL2VlZVlYsXscOXJEffv2Vd26dRUYGKhWrVppy5YtznFjjEaPHq2IiAgFBgYqISFB+/bts7Di8imp348//liJiYmqW7euHA6HduzYYV2xbnC1fgsLCzVixAi1atVKQUFBioyM1IMPPqijR49aXHX5lPQajx07Vs2aNVNQUJBq166thIQEbdy40cKKy6ekfi/22GOPyeFwaNq0aZVbpBuV1G///v0veS9PTk62sOLyK81rvGfPHt17770KCQlRUFCQOnTooEOHDllUcfmU1O+vX98Ly6uvvmph1eVTUs+nT59WSkqKGjZsqMDAQDVv3lyzZs2ysOLyKanfrKws9e/fX5GRkapRo4aSk5Or9L5HTEzMZX9nBw8eLEk6f/68Bg8erLp16+q6665Tjx49qvQ+Zkn9zp49W7fddpuCg4PlcDiUk5NjbcHldLV+T548qSeeeEJNmzZVYGCgoqOj9eSTTyo3N9fqssulpNf40Ucf1W9+8xsFBgaqfv366tq1q77//nuLqy4bQrsNtWjRQhkZGc7lq6++kiSdOXNGiYmJcjgcWr16tdavX6+CggJ16dJFxcXFFldddqdOnVLnzp3l6+urf//73/ruu+80ZcoU1a5d2zln0qRJev311zVr1ixt3LhRQUFBSkpK0vnz5y2svGxK0++ZM2d06623auLEiRZW6h4l9Xv27Flt27ZNo0aN0rZt2/Txxx8rPT1d9957r8WVl11pXuMbbrhBM2bM0DfffKOvvvpKMTExSkxM1LFjxyysvGxK0+8Fixcv1tdff63IyEgLKnWP0vabnJzs8l7+3nvvWVRx+ZWm5/379+vWW29Vs2bNtHbtWu3atUujRo1SQECAhZWXTWn6vfi1zcjI0FtvvSWHw6EePXpYWHnZlabnYcOGadmyZXrnnXe0Z88ePfXUU0pJSdGSJUssrLxsSurXGKNu3brpv//9rz799FNt375djRo1UkJCgs6cOWNx9WWzefNml9/ZFStWSJJ69uwpSRo6dKg+++wzffDBB1q3bp2OHj2q7t27W1lyuZTU79mzZ5WcnKxnn33WyjLd5mr9Hj16VEePHtXkyZO1e/duzZs3T8uWLdOAAQMsrrp8SnqN27Vrp9TUVO3Zs0fLly+XMUaJiYkqKiqysuyyMbCVMWPGmNatW192bPny5cbLy8vk5uY61+Xk5BiHw2FWrFhRSRW634gRI8ytt956xfHi4mITHh5uXn31Vee6nJwc4+/vb957773KKNGtSur3YgcOHDCSzPbt2yu2qAp0Lf1esGnTJiPJ/PjjjxVUVcUqS8+5ublGklm5cmUFVVVxStvv//7v/5oGDRqY3bt3m0aNGpnXXnut4ourAKXpt1+/fqZr166VU1AlKE3P9913n+nbt28lVVSxyvJvuGvXruaOO+6ooIoqXml6btGihRk3bpzLurZt25rnnnuuIkurECX1m56ebiSZ3bt3O9cVFRWZ+vXrmzlz5lRGiRVuyJAh5je/+Y0pLi42OTk5xtfX13zwwQfO8T179hhJJi0tzcIq3efifi+2Zs0aI8mcOnXKmsIqyJX6vWDRokXGz8/PFBYWVnJlFaeknnfu3GkkmR9++KGSKys/jrTb0L59+xQZGanGjRurT58+zlML8/Pz5XA45O/v75wbEBAgLy8v59H4qmjJkiVq3769evbsqdDQULVp00Zz5sxxjh84cECZmZlKSEhwrgsJCVHHjh2VlpZmRcnlUlK/1U1Z+s3NzZXD4VCtWrUqp0g3u9aeCwoKNHv2bIWEhKh169aVWKl7lKbf4uJiPfDAA3rmmWfUokULiyp1j9K+vmvXrlVoaKiaNm2qQYMG6cSJExZU6x4l9VxcXKylS5fqhhtuUFJSkkJDQ9WxY0d98skn1hVdDtf6bzgrK0tLly6t0ketStPzLbfcoiVLlujIkSMyxmjNmjXau3evEhMTLaq67ErqNz8/X5JczhTx8vKSv79/ld7nuqCgoEDvvPOOHn74YTkcDm3dulWFhYUu+1rNmjVTdHR0ldzX+rVf91vdlabf3NxcBQcHy8fHp5Krqxgl9XzmzBmlpqYqNjZWUVFRFlRYTlZ/agBX//rXv8yiRYvMzp07zbJly0x8fLyJjo42eXl5Jjs72wQHB5shQ4aYM2fOmNOnT5uUlBQjyQwcONDq0svM39/f+Pv7m5EjR5pt27aZf/zjHyYgIMDMmzfPGGPM+vXrjSRz9OhRl8f17NnT9OrVy4qSy6Wkfi9WHY60X0u/xhhz7tw507ZtW3P//fdXcqXuU9qeP/vsMxMUFGQcDoeJjIw0mzZtsqji8ilNvy+//LK56667nJ9+V+Uj7aXp97333jOffvqp2bVrl1m8eLGJi4szHTp0MD///LOFlZddST1nZGQYSaZGjRpm6tSpZvv27WbChAnG4XCYtWvXWlz9tbvW962JEyea2rVrm3PnzlVype5Tmp7Pnz9vHnzwQSPJ+Pj4GD8/PzN//nwLqy67kvotKCgw0dHRpmfPnubkyZMmPz/fvPLKK0aSSUxMtLj68nv//feNt7e3OXLkiDHGmHfffdf4+fldMq9Dhw5m+PDhlV2e2/2634tVxyPtV+vXGGOOHTtmoqOjzbPPPlvJlVWcK/U8c+ZMExQUZCSZpk2bVsmj7MYYQ2i3uVOnTpng4GDzP//zP8aYX06Rb9y4sXE4HMbb29v07dvXtG3b1jz22GMWV1p2vr6+Jj4+3mXdE088YTp16mSMqX6hvaR+L1YdQvu19FtQUGC6dOli2rRp43IZSFVT2p5Pnz5t9u3bZ9LS0szDDz9sYmJiTFZWVmWW6hYl9btlyxYTFhbm8h9pVQ7t1/I7fcH+/fur7OUPxpTc85EjR4wk8+c//9llTpcuXUzv3r0rrU53udbXuGnTpiYlJaUySqswpen51VdfNTfccINZsmSJ2blzp5k+fbq57rrrquQleqXpd8uWLaZ169ZGkvH29jZJSUnmD3/4g0lOTq7sct0uMTHR3HPPPc7vq3to/3W/F6uOof1q/ebm5pqbb77ZJCcnm4KCgkqurOJcqeecnByzd+9es27dOtOlSxfTtm3bKvkBK6fH21ytWrV0ww036IcffpAkJSYmav/+/crOztbx48f1z3/+U0eOHFHjxo0trrTsIiIi1Lx5c5d1cXFxzssCwsPDJemSO5hmZWU5x6qSkvqtbkrbb2FhoXr16qUff/xRK1asUHBwcGWW6Val7TkoKEjXX3+9OnXqpLlz58rHx0dz586tzFLdoqR+v/zyS2VnZys6Olo+Pj7y8fHRjz/+qKeffloxMTEWVFw+Zfk33LhxY9WrV8/5Xl7VlNRzvXr15OPjU23e267lNf7yyy+Vnp6uv/zlL5VVXoUoqedz587p2Wef1dSpU9WlSxfdeOONSklJ0X333afJkydbUXK5lOY1bteunXbs2KGcnBxlZGRo2bJlOnHiRJXe55KkH3/8UStXrnT5nQ0PD1dBQcEld1CvqvtaF7tcv9XZ1fr96aeflJycrJo1a2rx4sXy9fW1oEL3u1rPISEhatKkiX73u9/pww8/1Pfff6/FixdbUGX5ENpt7vTp09q/f78iIiJc1terV0+1atXS6tWrlZ2dXaXvtN25c2elp6e7rNu7d68aNWokSYqNjVV4eLhWrVrlHM/Ly9PGjRsVHx9fqbW6Q0n9Vjel6fdCYN+3b59WrlypunXrVnaZblXW17i4uNh5HWVVUlK/DzzwgHbt2qUdO3Y4l8jISD3zzDNavny5FSWXS1le3//93//ViRMnLnkvrypK6tnPz08dOnSoNu9t1/Iaz507V+3atauS96O4WEk9FxYWqrCwUF5erruO3t7eVfIv2FzLaxwSEqL69etr37592rJli7p27VpZZVaI1NRUhYaG6u6773aua9eunXx9fV32tdLT03Xo0KEqua91scv1W51dqd+8vDwlJibKz89PS5YsqZJ/2eNKSvsam1/OMq+S+1qcHm8zTz/9tFm7dq05cOCAWb9+vUlISDD16tUz2dnZxhhj3nrrLZOWlmZ++OEH889//tPUqVPHDBs2zOKqy2fTpk3Gx8fHvPTSS2bfvn3m3XffNTVq1DDvvPOOc84rr7xiatWq5bxGtGvXriY2NrZKnt5Smn5PnDhhtm/fbpYuXWokmYULF5rt27ebjIwMCysvm5L6LSgoMPfee69p2LCh2bFjh8nIyHAu+fn5FldfNiX1fPr0aTNy5EiTlpZmDh48aLZs2WIeeugh4+/v73Kn4qqiNL/Tv1aVT48vqd+ffvrJ/PWvfzVpaWnmwIEDZuXKlaZt27amSZMm5vz58xZXXzaleY0//vhj4+vra2bPnm327dtnpk+fbry9vc2XX35pYeVlU9rf6dzcXFOjRg3z5ptvWlSp+5Sm59///vemRYsWZs2aNea///2vSU1NNQEBAeaNN96wsPKyKU2/ixYtMmvWrDH79+83n3zyiWnUqJHp3r27hVWXX1FRkYmOjjYjRoy4ZOyxxx4z0dHRZvXq1WbLli0mPj7+kksIqpqr9ZuRkWG2b99u5syZYySZL774wmzfvt2cOHHCgkrd40r95ubmmo4dO5pWrVqZH374wWVfq6rea+WCK/W8f/9+8/LLL5stW7aYH3/80axfv9506dLF1KlTp0peikhot5n77rvPREREGD8/P9OgQQNz3333udwwYcSIESYsLMz4+vqaJk2amClTplzxzxpUJZ999plp2bKl8ff3N82aNTOzZ892GS8uLjajRo0yYWFhxt/f39x5550mPT3domrLr6R+U1NTjaRLljFjxlhTcDldrd8L1+1fblmzZo11RZfT1Xo+d+6c+dOf/mQiIyONn5+fiYiIMPfee2+VvRGdMSX/Tv9aVQ7txly937Nnz5rExERTv3594+vraxo1amQeeeQRk5mZaWHF5Vea13ju3Lnm+uuvNwEBAaZ169bmk08+saBS9yhNv//4xz9MYGCgycnJsaBC9yup54yMDNO/f38TGRlpAgICTNOmTav0fkhJ/f797383DRs2NL6+viY6Oto8//zzVfbD5AuWL19uJF12H+rcuXPm8ccfN7Vr1zY1atQwf/rTn6rkwYKLXa3fMWPGXHbfIzU1tfILdZMr9Xvhuv3LLQcOHLCmWDe5Us9Hjhwxf/jDH0xoaKjx9fU1DRs2NPfff7/5/vvvLaq0fBzGGFM5x/QBAAAAAMC14Jp2AAAAAABsitAOAAAAAIBNEdoBAAAAALApQjsAAAAAADZFaAcAAAAAwKYI7QAAAAAA2BShHQAAAAAAmyK0AwBggYMHD8rhcGjHjh2SpLVr18rhcCgnJ8fSuiDddttteuqpp6wuAwAASYR2AADcrn///nI4HHI4HPL19VVsbKyGDx+u8+fPO+dERUUpIyNDLVu2tLBS69ghGPNBCQCgKvCxugAAAKqj5ORkpaamqrCwUFu3blW/fv3kcDg0ceJESZK3t7fCw8MtrhIAANgdR9oBAKgA/v7+Cg8PV1RUlLp166aEhAStWLHCOf7r0+MvWL9+vW688UYFBASoU6dO2r17t8v4Rx99pBYtWsjf318xMTGaMmWKc+zCkeNfL/3799fBgwfl5eWlLVu2uGxv2rRpatSokYqLi52PX758udq0aaPAwEDdcccdys7O1r///W/FxcUpODhY999/v86ePevcRnFxsSZMmKDY2FgFBgaqdevW+vDDD8v18/vqq6/029/+VoGBgYqKitKTTz6pM2fOOMdjYmL08ssv6+GHH1bNmjUVHR2t2bNnu2xjw4YNuummmxQQEKD27dvrk08+cf7MDx48qNtvv12SVLt2befP6eKehg8frjp16ig8PFxjx44tVz8AAJQVoR0AgAq2e/dubdiwQX5+fiXOfeaZZzRlyhRt3rxZ9evXV5cuXVRYWChJ2rp1q3r16qXevXvrm2++0dixYzVq1CjNmzdPknTLLbcoIyPDuaxevVoBAQH63e9+p5iYGCUkJCg1NdXl+VJTU9W/f395ef3fLsHYsWM1Y8YMbdiwQYcPH1avXr00bdo0LViwQEuXLtV//vMfTZ8+3Tl/woQJevvttzVr1ix9++23Gjp0qPr27at169aV6ee1f/9+JScnq0ePHtq1a5fef/99ffXVV0pJSXGZN2XKFLVv317bt2/X448/rkGDBik9PV2SlJeXpy5duqhVq1batm2bxo8frxEjRjgfGxUVpY8++kiSlJ6eroyMDP397393js+fP19BQUHauHGjJk2apHHjxrl86AIAQKUxAADArfr162e8vb1NUFCQ8ff3N5KMl5eX+fDDD51zDhw4YCSZ7du3G2OMWbNmjZFkFi5c6Jxz4sQJExgYaN5//31jjDH333+/ueuuu1ye65lnnjHNmze/pIbjx4+bxo0bm8cff9y57v333ze1a9c258+fN8YYs3XrVuNwOMyBAwdcali5cqXzMRMmTDCSzP79+53rHn30UZOUlGSMMeb8+fOmRo0aZsOGDS7PP2DAAPPnP//5ij+j3//+92bIkCGXHRswYIAZOHCgy7ovv/zSeHl5mXPnzhljjGnUqJHp27evc7y4uNiEhoaaN9980xhjzJtvvmnq1q3rnG+MMXPmzLnsz/zUqVOX1Hbrrbe6rOvQoYMZMWLEFfsBAKCicKQdAIAKcPvtt2vHjh3auHGj+vXrp4ceekg9evQo8XHx8fHOr+vUqaOmTZtqz549kqQ9e/aoc+fOLvM7d+6sffv2qaioyLmusLBQPXr0UKNGjVyOHnfr1k3e3t5avHixJGnevHm6/fbbFRMT47LNG2+80fl1WFiYatSoocaNG7usy87OliT98MMPOnv2rO666y5dd911zuXtt9/W/v37S+z3cnbu3Kl58+a5bC8pKUnFxcU6cODAZet0OBwKDw931pWenu68zOCCm2++udQ1XLxtSYqIiHBuGwCAysSN6AAAqABBQUG6/vrrJUlvvfWWWrdurblz52rAgAEV/tyDBg3S4cOHtWnTJvn4/N9/9X5+fnrwwQeVmpqq7t27a8GCBS6h/gJfX1/n1xfugH8xh8Oh4uJiSdLp06clSUuXLlWDBg1c5vn7+5ep/tOnT+vRRx/Vk08+eclYdHT0Zev8dV3lVZHbBgDgWhDaAQCoYF5eXnr22Wc1bNgw3X///QoMDLzi3K+//toZTE+dOqW9e/cqLi5OkhQXF6f169e7zF+/fr1uuOEGeXt7S5KmTp2qRYsWacOGDapbt+4l2//LX/6ili1b6o033tDPP/+s7t27l6u35s2by9/fX4cOHdLvf//7cm3rgrZt2+q7775zfuhRFk2bNtU777yj/Px854cHmzdvdplz4R4DF5+lAACA3XB6PAAAlaBnz57y9vbWzJkzrzpv3LhxWrVqlXbv3q3+/furXr166tatmyTp6aef1qpVqzR+/Hjt3btX8+fP14wZM/TXv/5VkrRy5UoNHz5cr776qurVq6fMzExlZmYqNzfXuf24uDh16tRJI0aM0J///OerfoBQGjVr1tRf//pXDR06VPPnz9f+/fu1bds2TZ8+XfPnz7/qY48dO6YdO3a4LFlZWRoxYoQ2bNiglJQU7dixQ/v27dOnn356yY3orub+++9XcXGxBg4cqD179mj58uWaPHmypF+OmktSo0aN5HA49Pnnn+vYsWPOswYAALATQjsAAJXAx8dHKSkpmjRpksufLvu1V155RUOGDFG7du2UmZmpzz77zHlEuG3btlq0aJEWLlyoli1bavTo0Ro3bpzzT5V99dVXKioq0mOPPaaIiAjnMmTIEJfnGDBggAoKCvTwww+7pbfx48dr1KhRmjBhguLi4pScnKylS5cqNjb2qo9bsGCB2rRp47LMmTNHN954o9atW6e9e/fqt7/9rdq0aaPRo0crMjKy1DUFBwfrs88+044dO3TTTTfpueee0+jRoyXJeZ17gwYN9MILL+hvf/ubwsLCrulDAQAAKovDGGOsLgIAAFSe8ePH64MPPtCuXbusLqVSvfvuu3rooYeUm5tb7jMMAACoLFzTDgCAhzh9+rQOHjyoGTNm6MUXX7S6nAr39ttvq3HjxmrQoIF27typESNGqFevXgR2AECVwunxAAB4iJSUFLVr10633Xab206Nt7PMzEz17dtXcXFxGjp0qHr27KnZs2dbXRYAANeE0+MBAAAAALApjrQDAAAAAGBThHYAAAAAAGyK0A4AAAAAgE0R2gEAAAAAsClCOwAAAAAANkVoBwAAAADApgjtAAAAAADYFKEdAAAAAACbIrQDAAAAAGBT/w/3IrFlnWXYywAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 1200x300 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig,ax = plt.subplots(1,1,figsize=(12,3))\n",
    "sns.histplot([len(x) for x in ribozymes], binwidth=1)\n",
    "ax.set_xticks(np.arange(59,74))\n",
    "ax.set_xlabel('Ribozyme Length')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6232"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ribozymes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "plasmid = ''.join(rng.choice(dna) for _ in range(5000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5000"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(plasmid) # 5kbp long plasmid"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Review constraints and formulate gameplan"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, we have 6,232 simulated candidates, between 59 to 73bp in length and a plasmid of length 5kbp."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The design goal is straightforward. We have our simulated ribozyme library to start, and we know we want to use `EcoRI` and `AatII` as our restriction sites as part of the characterization process. What we can do now is use `Oligopool Calculator` to design the following.\n",
    "* Three primers, each 20bp in length,\n",
    "* The prefix and suffix barcodes each 11bp in length, and\n",
    "* Variable length spacers ranging from 3 to 17bp to pad the oligos to 170bp."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As constraints, we aim to have the following.\n",
    "* The primers should not have off-target binding within the library or to the backbone,\n",
    "* Addition of primers, barcodes or spacers should not introduce `EcoRI` and `AatII` motifs elsewhere, and\n",
    "* The barcodes should all have at least 3 mismatches separating each other."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Other realistic constraints, such as exclusion of all palindomic hexamers, or polymeric runs such as `'AAAAA'`, ..., `'CCCC'` may be included as they were in the original projects, but we will skip those for this demonstration. In reality, multiple constaints will depend on the design goals."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How do we begin? By asking `help` from `Oligopool Calculator`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package oligopool:\n",
      "\n",
      "NAME\n",
      "    oligopool\n",
      "\n",
      "DESCRIPTION\n",
      "    oligopool v2024.09.29\n",
      "    by ah\n",
      "\n",
      "    Automated design and analysis of oligopool libraries.\n",
      "\n",
      "    The various modules in Oligopool Calculator can be used\n",
      "    interactively in a jupyter notebook, or be used to define\n",
      "    scripts for design and analysis pipelines on the cloud.\n",
      "\n",
      "    Oligopool Calculator offers two modes of operation\n",
      "        -   Design Mode for designing oligopool library\n",
      "        - Analysis Mode for analyzing oligopool datasets\n",
      "\n",
      "    Design Mode workflow\n",
      "\n",
      "        1. Initialize a pandas DataFrame with core library elements\n",
      "            a. The DataFrame must contain a unique 'ID' column serving as primary key\n",
      "            b. All other columns in the DataFrame must be DNA sequences or a - (dash)\n",
      "        2. Next, define any optional background sequences via the background module\n",
      "        3. Add necessary oligopool elements with constraints via design modules\n",
      "        4. Optionally, split long oligos and pad them via assembly modules\n",
      "        5. Perform length checks as needed and finalize library via auxiliary modules\n",
      "\n",
      "        Example:\n",
      "        >>> import oligopool as op\n",
      "        >>>\n",
      "        >>> # Read initial library\n",
      "        >>> init_df = pd.read_csv('initial_library.csv')\n",
      "        >>>\n",
      "        >>> # Add oligo elements\n",
      "        >>> barcode_df, barcod_stats = op.barcode(input_data=init_df, ...)\n",
      "        >>> primer_df,  primer_stats = op.primer(input_data=barcode_df, ...)\n",
      "        >>>\n",
      "        >>> # Check length statistics\n",
      "        >>> op.lenstat(input_data=primer_df)\n",
      "        >>>\n",
      "        >>> # Split and pad longer oligos if needed\n",
      "        >>> split_df, split_stats = op.split(input_data=primer_df, ...)\n",
      "        >>> pad_df, pad_stats     = op.pad(input_data=split_df, ...)\n",
      "        >>>\n",
      "        >>> # Finalize and export library\n",
      "        >>> final_df, final_stats = op.final(input_data=split_df, ...)\n",
      "        >>> final_df.to_csv('final_library.csv')\n",
      "\n",
      "        Background module available\n",
      "            - background\n",
      "\n",
      "        Design modules available\n",
      "            - primer\n",
      "            - barcode\n",
      "            - spacer\n",
      "            - motif\n",
      "\n",
      "        Assembly modules available\n",
      "            - split\n",
      "            - pad\n",
      "\n",
      "        Auxiliary modules available\n",
      "            - lenstat\n",
      "            - final\n",
      "\n",
      "    Analysis Mode workflow\n",
      "\n",
      "        1. Index one or more CSVs containing the barcode and anchor information\n",
      "        2. Pack all NGS FastQ files, optionally merging them if required\n",
      "        3. If a barcode and its association with the core variant is to be counted use acount\n",
      "        4. If multiple barcode combinations are to be counted use xcount\n",
      "        5. Combine count matrices and perform stats and ML as necessary\n",
      "\n",
      "        Example:\n",
      "        >>> import oligopool as op\n",
      "        >>>\n",
      "        >>> # Read marked up library\n",
      "        >>> final_df = pd.read_csv('final_library.csv')\n",
      "        >>>\n",
      "        >>> # Index the barcodes and save the indexes\n",
      "        >>> bc1_index_stats = op.index(input_data=final_df, barcode_column='BC1', ...)\n",
      "        >>> bc2_index_stats = op.index(input_data=final_df, barcode_column='BC2', ...)\n",
      "        >>>\n",
      "        >>> # Pack the FastQ files\n",
      "        >>> sam1_pack_stats = op.pack(r1_file='sample_1_R1.fq.gz', ...)\n",
      "        >>> sam2_pack_stats = op.pack(r1_file='sample_2_R1.fq.gz', ...)\n",
      "        >>>\n",
      "        >>> # Compute and write the barcode combination count matrix\n",
      "        >>> xount_stats = op.xcount(index_files=['bc1_index', 'bc2_index'],\n",
      "        ...                         pack_file='sample_1_pack', ...)\n",
      "        >>>\n",
      "        >>> # Read the count matrix and continue analysis\n",
      "        >>> count_df = pd.read_csv('library_count_matrix')\n",
      "\n",
      "        Indexing module available\n",
      "            - index\n",
      "\n",
      "        Packing module available\n",
      "            - pack\n",
      "\n",
      "        Counting modules available\n",
      "            - acount\n",
      "            - xcount\n",
      "\n",
      "    You can learn more about each module using help.\n",
      "    >>> import oligopool as op\n",
      "    >>>\n",
      "    >>> help(op)\n",
      "    >>> help(op.primer)\n",
      "    >>> help(op.barcode)\n",
      "    ...\n",
      "    >>> help(op.xcount)\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    acount\n",
      "    background\n",
      "    barcode\n",
      "    final\n",
      "    index\n",
      "    lenstat\n",
      "    motif\n",
      "    pack\n",
      "    pad\n",
      "    primer\n",
      "    spacer\n",
      "    split\n",
      "    xcount\n",
      "\n",
      "SUBMODULES\n",
      "    base\n",
      "\n",
      "VERSION\n",
      "    2024.09.29\n",
      "\n",
      "AUTHOR\n",
      "    Ayaan Hossain\n",
      "\n",
      "FILE\n",
      "    /home/ec2-user/dev/oligopool/oligopool/__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(op)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To begin, per the help, will need to use the `barcode`, `primer`, `motif` and `spacer` modules to add the library elements."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Primer design review"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generally, the primers for a library are the most constrained elements in the library because they have to be very specific to the library with minimal off-targeting. For oligopool libraries, it is recommended to design the primers first and get them out of the way. At any step in the design process if we get stuck, `Oligopool Calculator` will tell us what is preventing it from succeeding and we can relax or modify any constraints accordingly."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How do we design primers? Let's ask for help!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function primer in module oligopool.primer:\n",
      "\n",
      "primer(input_data: str | pandas.core.frame.DataFrame, oligo_length_limit: int, primer_sequence_constraint: str, primer_type: int, minimum_melting_temperature: float, maximum_melting_temperature: float, maximum_repeat_length: int, primer_column: str, output_file: str | None = None, paired_primer_column: str | None = None, left_context_column: str | None = None, right_context_column: str | None = None, excluded_motifs: list | str | pandas.core.frame.DataFrame | None = None, background_directory: str | None = None, verbose: bool = True) -> Tuple[pandas.core.frame.DataFrame, dict]\n",
      "    Designs constrained primers with specified melting temperature and non-repetitiveness\n",
      "    for all variants in the oligopool. Ensures compatibility with paired primers and minimizes\n",
      "    dimer formation. Returns a DataFrame of designed primers, optionally saving to <output_file>\n",
      "    in CSV format.\n",
      "\n",
      "    Required Parameters:\n",
      "        - input_data (str / pd.DataFrame): Path to a CSV file or DataFrame with annotated oligopool variants.\n",
      "        - oligo_length_limit (int): Maximum allowed oligo length (≥ 4).\n",
      "        - primer_sequence_constraint (int): IUPAC degenerate sequence constraint.\n",
      "        - primer_type (int): Primer type (0 for forward, 1 for reverse).\n",
      "        - minimum_melting_temperature (float): Minimum Tm (≥ 25°C).\n",
      "        - maximum_melting_temperature (float): Maximum Tm (≤ 95°C).\n",
      "        - maximum_repeat_length (int): Max shared repeat length with oligos (≥ 6).\n",
      "        - primer_column (str): Column name for the designed primer.\n",
      "\n",
      "    Optional Parameters:\n",
      "        - output_file (str / None): Filename for output DataFrame (default: None).\n",
      "        - paired_primer_column (str / None): Column for paired primer sequence (default: None).\n",
      "        - left_context_column (str / None): Column for left DNA context (default: None).\n",
      "        - right_context_column (str / None): Column for right DNA context (default: None).\n",
      "        - excluded_motifs (list / str / pd.DataFrame / None): Motifs to exclude;\n",
      "            can be a CSV path or DataFrame (default: None).\n",
      "        - background_directory (str / None): Directory for background k-mer sequences (default: None).\n",
      "        - verbose (bool): If True, logs updates to stdout (default: True).\n",
      "\n",
      "    Returns:\n",
      "        - A pandas DataFrame of designed primers; saves to <output_file> if specified.\n",
      "        - A dictionary of stats from pipeline steps.\n",
      "\n",
      "    Notes:\n",
      "        - <input_data> must contain a unique 'ID' column.\n",
      "        - All other columns be non-empty DNA strings or a single dash ('-').\n",
      "        - Column names in <input_data> must be unique, excluding <primer_column>.\n",
      "        - At least one of <left_context_column> or <right_context_column> must be specified.\n",
      "        - The paired primer type is inferred based on the current primer type.\n",
      "        - If a paired primer is specified, Tm of the designed primer is optimized within 1°C of it.\n",
      "        - <maximum_repeat_length> controls non-repetitiveness against <input_data> only, not <background>.\n",
      "        - If <excluded_motifs> is a CSV or DataFrame, it must have 'ID' and 'Exmotif' columns.\n",
      "        - Constant motifs in sequence constraint may lead to sub-optimal primers.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(op.primer)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Two things should stand out here. First, the primer design involves consideration of background sequences to minimize off-target amplification, such as from the plasmid we set up. Second, primer design involves consideration of pairing information -- this implies primers need to be designed in pairs in a particular order. If we only had two primers (one forward and one reverse) the order would not matter. But in our problem, there are two forward primers (pink and orange) for both of which we have a single reverse primer (yellow)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us first solve the background problem."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Storing the plasmid background"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Oligopool Calculator` makes it easy to define a background for primer design via the `background` module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function background in module oligopool.background:\n",
      "\n",
      "background(input_data: list | str | pandas.core.frame.DataFrame, maximum_repeat_length: int, output_directory: str, verbose=True) -> dict\n",
      "    The background function creates a k-mer database from a list or CSV file of background sequences.\n",
      "    This database is used during primer design to ensure primers are non-repetitive to the background,\n",
      "    minimizing off-target amplification. Non-repetitiveness is regulated by the maximum shared repeat\n",
      "    length parameter. The generated database is saved in the specified <output_directory>.\n",
      "\n",
      "    Required Parameters:\n",
      "        - input_data (list / str / pd.DataFrame): background for primers; can be a CSV file or a DataFrame.\n",
      "        - maximum_repeat_length (int): Max repeat length between primers and background (between 6 and 20).\n",
      "        - output_directory (str): Directory to store the generated background k-mer database.\n",
      "        - verbose (bool): If True, logs updates to stdout (default: True).\n",
      "\n",
      "    Returns:\n",
      "        - A dictionary of stats from pipeline steps.\n",
      "\n",
      "    Notes:\n",
      "        - If <input_data> is a CSV or DataFrame, must contain 'ID' and a 'Sequence' column with DNA strings.\n",
      "        - <maximum_repeat_length> here controls non-repetitiveness of primers to <background> only.\n",
      "        - For manipulation, use vectorDB to operate on the background (see help(oligopool.vectorDB)).\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(op.background)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> **Note:** If you want to modify the background post-hoc, or want to define your background using a custom strategy, directly use the `vectorDB` module, which behaves like a dictionary. For details use `help(op.vectorDB)`. The `background` module just wraps around it and offers an one-time use simple interface to `vectorDB`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "background_directory = 'demo.oligopool.background'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove the background if one exists already -- for demo\n",
    "if os.path.exists(background_directory):\n",
    "    op.vectorDB(background_directory, maxreplen=11, mode=1).drop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[Oligopool Calculator: Design Mode - Background]\n",
      "                                                  \n",
      " Required Arguments\n",
      " Background      Data: 1 Unique Sequence(s)\n",
      "    Maximum    Repeat: Up to 10 Base Pair(s) Background Repeats\n",
      "     Output Directory: demo.oligopool.background                \n",
      "                                                 \n",
      "[Computing Background]\n",
      " Sequence 1: TGTGCCAACCGACTTCTGAA... Inserted\n",
      " Time Elapsed: 0.02 sec                       \n",
      "                        \n",
      "[Background Statistics]\n",
      " Background Status: Successful\n",
      "      k-mer  Space: 2,097,152 Unique 11-mers\n",
      "       Fill  Count:     4,973 Unique 11-mers (  0.24 %)\n",
      "       Left  Count: 2,092,179 Unique 11-mers ( 99.76 %) \n",
      " Time Elapsed: 0.03 sec                                 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "background_stats = op.background(\n",
    "    input_data=plasmid,\n",
    "    maximum_repeat_length=10,\n",
    "    output_directory=background_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'status': True,\n",
       " 'basis': 'solved',\n",
       " 'step': 1,\n",
       " 'step_name': 'computing-background',\n",
       " 'vars': {'kmer_space': 2097152, 'fill_count': 4973, 'left_count': 2092179},\n",
       " 'warns': {}}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "background_stats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We've now stored the background, and are ready to design the three primers."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> **Note**: All `Oligopool Calculator` functions return a pipeline statistics dictionary recapitulating the steps performed in the function."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Designing the three primers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All `Design Mode` input and output is chained via a `pandas` `DataFrame` to facilitate piplined designs. Let us store the ribozyme candidates in a dataframe and start the design process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "init_df = pd.DataFrame({\n",
    "    'ID': [f'ribozyme_{x+1:06d}' for x in range(len(ribozymes))],\n",
    "    'EcoRI': 'GAATTC',\n",
    "    'Variant':ribozymes,\n",
    "    'AatII': 'GACGTC',\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>EcoRI</th>\n",
       "      <th>Variant</th>\n",
       "      <th>AatII</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ribozyme_000001</td>\n",
       "      <td>GAATTC</td>\n",
       "      <td>CGTTCAGAAGCGCGCGACTGTACCGTCGTTAAGCACCTGACGTACT...</td>\n",
       "      <td>GACGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ribozyme_000002</td>\n",
       "      <td>GAATTC</td>\n",
       "      <td>TCTTCTATCATAGTTAGGCTACGGATCCTCCTCTAGGACACACCCG...</td>\n",
       "      <td>GACGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ribozyme_000003</td>\n",
       "      <td>GAATTC</td>\n",
       "      <td>GCTGATTCATTCCAACACTCTTGAGGCCGTTTCTATAAACGGTGAC...</td>\n",
       "      <td>GACGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ribozyme_000004</td>\n",
       "      <td>GAATTC</td>\n",
       "      <td>CATATCTCTGTTCCTCTGCTGTAATCATGAGTCGGATCACTGGTTG...</td>\n",
       "      <td>GACGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ribozyme_000005</td>\n",
       "      <td>GAATTC</td>\n",
       "      <td>GCATCCGTTGGTAACTTTCAATTGTTCAGGTCTCTGGCTTATCACG...</td>\n",
       "      <td>GACGTC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                ID   EcoRI                                            Variant  \\\n",
       "0  ribozyme_000001  GAATTC  CGTTCAGAAGCGCGCGACTGTACCGTCGTTAAGCACCTGACGTACT...   \n",
       "1  ribozyme_000002  GAATTC  TCTTCTATCATAGTTAGGCTACGGATCCTCCTCTAGGACACACCCG...   \n",
       "2  ribozyme_000003  GAATTC  GCTGATTCATTCCAACACTCTTGAGGCCGTTTCTATAAACGGTGAC...   \n",
       "3  ribozyme_000004  GAATTC  CATATCTCTGTTCCTCTGCTGTAATCATGAGTCGGATCACTGGTTG...   \n",
       "4  ribozyme_000005  GAATTC  GCATCCGTTGGTAACTTTCAATTGTTCAGGTCTCTGGCTTATCACG...   \n",
       "\n",
       "    AatII  \n",
       "0  GACGTC  \n",
       "1  GACGTC  \n",
       "2  GACGTC  \n",
       "3  GACGTC  \n",
       "4  GACGTC  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "init_df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The optimal solution here is to design the innermost (pink) forward primer first, and then design the reverse primer (yellow), followed by designing the outermost (orange) forward primer. This is allows all three primers to be compatible with each other."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# help(op.primer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[Oligopool Calculator: Design Mode - Primer]\n",
      "                                              \n",
      " Required Arguments\n",
      "      Input Data       : DataFrame w/ 6,232 Record(s)\n",
      "      Oligo Limit      : At most 170 Base Pair(s)     \n",
      "     Primer Sequence   : A 20 Base Pair IUPAC Constraint\n",
      "     Primer Type       : Forward Primer Design           \n",
      "    Melting Temperature: 53 to 55 °C           \n",
      "     Repeat Length     : Up to 8 Base Pair(s) Oligopool Repeats\n",
      "     Primer Column     : Output in Column 'PinkForward'         \n",
      "     Output File       : In-Memory DataFrame            \n",
      "                                             \n",
      " Optional Arguments\n",
      "     Paired Primer     : None Specified\n",
      "       Left Context    : Input from Column 'Variant'\n",
      "      Right Context    : None Specified              \n",
      "   Excluded Motifs     : 2 Unique Motif(s)\n",
      " Background Database   : Contains 4,973 Unique 11-mers\n",
      "                                                       \n",
      "[Step 1: Parsing Oligo Limit]\n",
      " Maximum Oligo Length:  170 Base Pair(s)\n",
      " Input Variant Length:   71 to 85 Base Pair(s)\n",
      "  Reqd. Primer Length:   20 Base Pair(s)       \n",
      " Free Space Available:   85 to 99 Base Pair(s)\n",
      " Time Elapsed: 0.01 sec                        \n",
      " Verdict: Primer Design Possibly Feasible\n",
      "                                          \n",
      "[Step 2: Parsing Excluded Motifs]\n",
      " Sorted and Enqued: 2 Unique Excluded Motif(s)\n",
      " Excluded Motif Length Distribution              \n",
      "   - 2 Motif(s) of Length 6         \n",
      " Time Elapsed: 0.00 sec                         \n",
      " Verdict: Primer Design Possibly Feasible\n",
      "                                          \n",
      "[Step 3: Parsing Primer Sequence]\n",
      " Design Space: 274,877,906,944 Possible Primer(s)\n",
      " Found 0 Excluded Motif(s)                        \n",
      " Found 0 Internal Repeat(s)             \n",
      " Found 0 Palindrome(s)             \n",
      " Found 0 Paired Repeat(s)\n",
      " Time Elapsed: 0.00 sec   \n",
      " Verdict: Primer Design Possibly Feasible\n",
      "                                          \n",
      "[Step 4: Parsing Melting Temperature]\n",
      "  Parameter Source: From Input Argument(s)\n",
      " Required Tm Range: 53.00 to 55.00 °C      \n",
      " Possible Tm Range: 22.62 to 71.50 °C             \n",
      "  Updated Tm Range: 53.00 to 55.00 °C \n",
      " Found 0 Melting Temperature Conflict(s)\n",
      " Time Elapsed: 2.25 sec                  \n",
      " Verdict: Primer Design Possibly Feasible\n",
      "                                          \n",
      "[Step 5: Extracting Context Sequences]\n",
      "  Left Context: 3,179 Unique Sequence(s)           \n",
      " Right Context:     0 Unique Sequence(s) \n",
      "  Time Elapsed: 0.00 sec                 \n",
      "                         \n",
      "[Step 6: Parsing Edge Effects]\n",
      "  Left Context 3,179: ...GCCGTT Prevents 1 Prefix(es)\n",
      " Time Elapsed: 0.87 sec        \n",
      " Verdict: Primer Design Possibly Feasible\n",
      "                                          \n",
      "[Step 7: Parsing Oligopool Repeats]\n",
      " Repeat Source: From Entire Oligopool\n",
      "  k-mer Space : 131,072 Unique 9-mers \n",
      "   Fill Count : 124,619 Unique 9-mers ( 95.08 %)\n",
      "   Free Count :   6,453 Unique 9-mers (  4.92 %) \n",
      " Verdict: Primer Design Possibly Feasible        \n",
      " Time Elapsed: 1.65 sec                   \n",
      "                        \n",
      "[Step 8: Computing Primer]\n",
      " Candidate: Primer GTACCCCACCTATCCGCCTT is Accepted                           \n",
      " Time Elapsed: 0.65 sec\n",
      "                        \n",
      "[Primer Design Statistics]\n",
      "      Design Status     : Successful\n",
      "     Melting Temperature:  53.02 °C  \n",
      "          GC Content    :  60.00 %  \n",
      "     Hairpin MFE        :   0.00 kcal/mol\n",
      "   Homodimer MFE        :  -5.67 kcal/mol \n",
      " Time Elapsed: 5.44 sec                   \n",
      "\n"
     ]
    }
   ],
   "source": [
    "pink_primer_df, pink_primer_stats = op.primer(\n",
    "    input_data=init_df,\n",
    "    oligo_length_limit=170,                    # Must not exceed 170bp at any stage\n",
    "    primer_sequence_constraint='N'*18 + 'WW',  # The last two bases must be weak\n",
    "    primer_type=0,                             # This is a forward primer\n",
    "    paired_primer_column=None,                 # No paired primer yet\n",
    "    minimum_melting_temperature=53,            # Tm = 53°C at least\n",
    "    maximum_melting_temperature=55,            # Tm = 55°C at most\n",
    "    maximum_repeat_length=8,                  # No more than 14-mer repeats\n",
    "    primer_column='PinkForward',               # Store designs in 'PinkForward' column\n",
    "    left_context_column='Variant',             # Next to 'Variant' column\n",
    "    right_context_column=None,                 # Before 'AatII' column\n",
    "    excluded_motifs=['GAATTC', 'GACGTC'],      # Ensure EcoRI and AatII do not emerge as motifs\n",
    "    background_directory=background_directory, # As setup before\n",
    "    output_file=None,                          # In-memory return, no CSV written here\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us now design the two remaining primers in silent mode."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "yellow_primer_df, _ = op.primer(\n",
    "    input_data=pink_primer_df,                 # Updated DataFrame along pipeline chain\n",
    "    oligo_length_limit=170,\n",
    "    primer_sequence_constraint='WW' + 'N'*18,  # Last two bases must be weak in reverse complement\n",
    "    primer_type=1,                             # This is a reverse primer\n",
    "    paired_primer_column='PinkForward',        # Yellow reverse primer paired with pink forward primer\n",
    "    minimum_melting_temperature=53,\n",
    "    maximum_melting_temperature=55,\n",
    "    maximum_repeat_length=8,\n",
    "    primer_column='YellowReverse',             # Store designs in 'YellowReverse' column\n",
    "    left_context_column='PinkForward',         # Next to 'PinkForward' column\n",
    "    right_context_column='AatII',              # Before 'AatII' column\n",
    "    excluded_motifs=['GAATTC', 'GACGTC'],\n",
    "    background_directory=background_directory,\n",
    "    output_file=None,                          # In-memory return, no CSV written here\n",
    "    verbose=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "orange_primer_df, _ = op.primer(\n",
    "    input_data=yellow_primer_df,               # Updated DataFrame along pipeline chain\n",
    "    oligo_length_limit=170,\n",
    "    primer_sequence_constraint='N'*18 + 'WW',  # The last two bases must be weak\n",
    "    primer_type=0,                             # This is a forward primer\n",
    "    paired_primer_column='YellowReverse',      # Orange forward primer paired with yellow reverse primer\n",
    "    minimum_melting_temperature=53,\n",
    "    maximum_melting_temperature=55,\n",
    "    maximum_repeat_length=8,\n",
    "    primer_column='OrangeForward',             # Store designs in 'OrangeForward' column\n",
    "    left_context_column='EcoRI',               # Next to 'EcoRI' column\n",
    "    right_context_column='Variant',            # Before 'AatII' column\n",
    "    excluded_motifs=['GAATTC', 'GACGTC'],\n",
    "    background_directory=background_directory,\n",
    "    output_file=None,                          # In-memory return, no CSV written here\n",
    "    verbose=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EcoRI</th>\n",
       "      <th>OrangeForward</th>\n",
       "      <th>Variant</th>\n",
       "      <th>PinkForward</th>\n",
       "      <th>YellowReverse</th>\n",
       "      <th>AatII</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ribozyme_000001</th>\n",
       "      <td>GAATTC</td>\n",
       "      <td>CCCTCCATCGCTTCCTGTTT</td>\n",
       "      <td>CGTTCAGAAGCGCGCGACTGTACCGTCGTTAAGCACCTGACGTACT...</td>\n",
       "      <td>GTACCCCACCTATCCGCCTT</td>\n",
       "      <td>AAGTCGTCGGTAATCGGGGT</td>\n",
       "      <td>GACGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ribozyme_000002</th>\n",
       "      <td>GAATTC</td>\n",
       "      <td>CCCTCCATCGCTTCCTGTTT</td>\n",
       "      <td>TCTTCTATCATAGTTAGGCTACGGATCCTCCTCTAGGACACACCCG...</td>\n",
       "      <td>GTACCCCACCTATCCGCCTT</td>\n",
       "      <td>AAGTCGTCGGTAATCGGGGT</td>\n",
       "      <td>GACGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ribozyme_000003</th>\n",
       "      <td>GAATTC</td>\n",
       "      <td>CCCTCCATCGCTTCCTGTTT</td>\n",
       "      <td>GCTGATTCATTCCAACACTCTTGAGGCCGTTTCTATAAACGGTGAC...</td>\n",
       "      <td>GTACCCCACCTATCCGCCTT</td>\n",
       "      <td>AAGTCGTCGGTAATCGGGGT</td>\n",
       "      <td>GACGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ribozyme_000004</th>\n",
       "      <td>GAATTC</td>\n",
       "      <td>CCCTCCATCGCTTCCTGTTT</td>\n",
       "      <td>CATATCTCTGTTCCTCTGCTGTAATCATGAGTCGGATCACTGGTTG...</td>\n",
       "      <td>GTACCCCACCTATCCGCCTT</td>\n",
       "      <td>AAGTCGTCGGTAATCGGGGT</td>\n",
       "      <td>GACGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ribozyme_000005</th>\n",
       "      <td>GAATTC</td>\n",
       "      <td>CCCTCCATCGCTTCCTGTTT</td>\n",
       "      <td>GCATCCGTTGGTAACTTTCAATTGTTCAGGTCTCTGGCTTATCACG...</td>\n",
       "      <td>GTACCCCACCTATCCGCCTT</td>\n",
       "      <td>AAGTCGTCGGTAATCGGGGT</td>\n",
       "      <td>GACGTC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  EcoRI         OrangeForward  \\\n",
       "ID                                              \n",
       "ribozyme_000001  GAATTC  CCCTCCATCGCTTCCTGTTT   \n",
       "ribozyme_000002  GAATTC  CCCTCCATCGCTTCCTGTTT   \n",
       "ribozyme_000003  GAATTC  CCCTCCATCGCTTCCTGTTT   \n",
       "ribozyme_000004  GAATTC  CCCTCCATCGCTTCCTGTTT   \n",
       "ribozyme_000005  GAATTC  CCCTCCATCGCTTCCTGTTT   \n",
       "\n",
       "                                                           Variant  \\\n",
       "ID                                                                   \n",
       "ribozyme_000001  CGTTCAGAAGCGCGCGACTGTACCGTCGTTAAGCACCTGACGTACT...   \n",
       "ribozyme_000002  TCTTCTATCATAGTTAGGCTACGGATCCTCCTCTAGGACACACCCG...   \n",
       "ribozyme_000003  GCTGATTCATTCCAACACTCTTGAGGCCGTTTCTATAAACGGTGAC...   \n",
       "ribozyme_000004  CATATCTCTGTTCCTCTGCTGTAATCATGAGTCGGATCACTGGTTG...   \n",
       "ribozyme_000005  GCATCCGTTGGTAACTTTCAATTGTTCAGGTCTCTGGCTTATCACG...   \n",
       "\n",
       "                          PinkForward         YellowReverse   AatII  \n",
       "ID                                                                   \n",
       "ribozyme_000001  GTACCCCACCTATCCGCCTT  AAGTCGTCGGTAATCGGGGT  GACGTC  \n",
       "ribozyme_000002  GTACCCCACCTATCCGCCTT  AAGTCGTCGGTAATCGGGGT  GACGTC  \n",
       "ribozyme_000003  GTACCCCACCTATCCGCCTT  AAGTCGTCGGTAATCGGGGT  GACGTC  \n",
       "ribozyme_000004  GTACCCCACCTATCCGCCTT  AAGTCGTCGGTAATCGGGGT  GACGTC  \n",
       "ribozyme_000005  GTACCCCACCTATCCGCCTT  AAGTCGTCGGTAATCGGGGT  GACGTC  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orange_primer_df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We are done with primers. Barcodes up next!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
